src/java.base/share/classes/java/util/LinkedHashMap.java

```diff
@@ -528,40 +601,63 @@ protected boolean removeEldestEntry(Map.Entry<K,V> eldest) {
      *
      * @return a set view of the keys contained in this map
      */
-    public Set<K> keySet() {
-        Set<K> ks = keySet;
+    public SequencedSet<K> keySet() {
```

Comment:
Changing the return type means that subclasses of `LinkedHashMap` that override `keySet()` will no longer compile, unless they also change the return type.

Reply:
Yes. Previously compiled binaries will also throw `NoSuchMethodError` if they call `super.keySet()`. Clearly these are incompatibilities; the question is whether they are severe enough to warrant pursuing a different approach. If you have any thoughts about how to assess the severity of this change, please let me know.

Reply:
Won’t `super.keySet()` resolve to a bridge method in that case?

Reply:
Oh yes, in this case it might resolve to a bridge method, and it might actually work. In this case. However, the compatibility matrix for covariant overrides and bridge methods is quite complex; see

https://mail.openjdk.java.net/pipermail/core-libs-dev/2012-February/009119.html

for why we decided not to add covariant `clone` overrides. Similar issues might arise here.

Reply:
For what it's worth, I've done a detailed compatibility analysis of the cases here, and I've decided that adding covariant overrides for keySet/values/entrySet is too dangerous. See the full analysis in the documents attached to the CSR request:

https://bugs.openjdk.org/browse/JDK-8266572